detector_test.go

```diff
@@ -591,7 +592,840 @@ func testDetector(t *testing.T, when spec.G, it spec.S) {
+					Group: []lifecycle.Buildpack{
+						{ID: "A", Version: "v1", API: "0.3"},
+						{ID: "B", Version: "v1", API: "0.2"},
+					},
+				}); s != "" {
+					t.Fatalf("Unexpected group:\n%s\n", s)
+				}
+
+				if !hasEntries(dr.BuildPlan.Entries, []lifecycle.BuildPlanEntry{
+					{
+						Providers: []lifecycle.Buildpack{
+							{ID: "X", Version: "1.0.0", Privileged: true},
+							{ID: "A", Version: "v1"},
+						},
+						Requires: []lifecycle.Require{{Name: "dep1", Mixin: false}},
+					},
+				}) {
+					t.Fatalf("Unexpected build entries:\n%+v\n", dr.BuildPlan.Entries)
+				}
+
+				if s := allLogs(logHandler); !strings.HasSuffix(s,
+					"======== Results ========\n"+
+						"pass: X@1.0.0\n"+
+						"pass: A@v1\n"+
+						"pass: B@v1\n"+
+						"Resolving plan... (try #1)\n"+
+						"skip: X@1.0.0[run] not required\n"+
+						"X 1.0.0\n"+
+						"A v1\n"+
+						"B v1\n",
+				) {
+					t.Fatalf("Unexpected log:\n%s\n", s)
+				}
+			})
+
+			it("should return a build plan with matched dependencies from privileged buildpacks", func() {
+				toappfile("\n[[provides]]\n name = \"dep1\"", "detect-plan-X-1.0.0.toml")
+				toappfile("\n[[requires]]\n name = \"dep1\"", "detect-plan-B-v1.toml")
+
+				dr, err := lifecycle.BuildpackOrder{
+					{Group: []lifecycle.Buildpack{
+						{ID: "X", Version: "1.0.0", Privileged: true},
+						{ID: "B", Version: "v1"},
+					}},
+				}.Detect(config)
+				if err != nil {
+					t.Fatalf("Unexpected error:\n%s\n", err)
+				}
+
+				if s := cmp.Diff(dr.BuildPrivilegedGroup, lifecycle.BuildpackGroup{
+					Group: []lifecycle.Buildpack{
+						{ID: "X", Version: "1.0.0", API: "0.3", Privileged: true},
+					},
+				}); s != "" {
+					t.Fatalf("Unexpected priv group:\n%s\n", s)
+				}
+
+				if s := cmp.Diff(dr.BuildGroup, lifecycle.BuildpackGroup{
+					Group: []lifecycle.Buildpack{
+						{ID: "B", Version: "v1", API: "0.2"},
+					},
+				}); s != "" {
+					t.Fatalf("Unexpected group:\n%s\n", s)
+				}
+
+				// since it is not a mixin, dep1 is really build:dep1
+				if s := cmp.Diff(dr.RunGroup, lifecycle.BuildpackGroup{
+					Group: []lifecycle.Buildpack{},
+				}); s != "" {
+					t.Fatalf("Unexpected run group:\n%s\n", s)
+				}
+
+				if !hasEntries(dr.BuildPlan.Entries, []lifecycle.BuildPlanEntry{
+					{
+						Providers: []lifecycle.Buildpack{
+							{ID: "X", Version: "1.0.0", Privileged: true},
+						},
+						Requires: []lifecycle.Require{{Name: "dep1", Mixin: false}},
+					},
+				}) {
+					t.Fatalf("Unexpected entries:\n%+v\n", dr.BuildPlan.Entries)
+				}
+
+				if s := allLogs(logHandler); !strings.HasSuffix(s,
+					"======== Results ========\n"+
+						"pass: X@1.0.0\n"+
+						"pass: B@v1\n"+
+						"Resolving plan... (try #1)\n"+
+						"skip: X@1.0.0[run] not required\n"+
+						"X 1.0.0\n"+
+						"B v1\n",
+				) {
+					t.Fatalf("Unexpected log:\n%s\n", s)
+				}
+			})
+
+			it("should fail if all requires are not met by mixins", func() {
```

Comment:
Today, static mixins are _only_ validated by `pack` (and other platforms, I suppose). I do think lifecycle needs to do this, but I wasn't ready to include that as part of this. Lifecycle currently only parses that `stack.toml` to figure out the run image and mirrors IIRC.

Reply:
Just summarizing here in case it would help others - 

Eventually, every combination of static (via buildpack.toml) or dynamic (via the build plan) "requires" and static (via the stack) or dynamic (via stack buildpacks) "provides" will be supported. Today, we have static -> static support with `pack` validating mixins up front. This PR introduces support for dynamic -> dynamic. The others will be supported in the future.